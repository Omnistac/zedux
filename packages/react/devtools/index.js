if (process.env.NODE_ENV !== 'development') {
  module.exports = {
    getStateHub: function getStateHub() {},
    StateHub: function StateHub() {
      return null
    },
  }
} else {
  module.exports = require('./dev')
}
