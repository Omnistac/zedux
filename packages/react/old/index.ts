export * from './AppProvider'
export { globalStore } from './global/store'
export * from './createAtom'
export * from './createMolecule'
export * from './types'
export * from './useAtom'
export * from './useMolecule'
export * from './useStore'
