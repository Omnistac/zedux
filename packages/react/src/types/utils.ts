import { AtomTemplateBase } from '../classes/templates/AtomTemplateBase'
import { AtomInstance } from '../classes/instances/AtomInstance'
import { AtomInstanceBase } from '../classes/instances/AtomInstanceBase'

export type AnyAtomInstance = AtomInstance<any>
export type AnyAtomInstanceBase = AtomInstanceBase<
  any,
  AtomTemplateBase<any, AtomInstance<any>>
>
export type AnyAtomTemplate = AtomTemplateBase<any, any>
