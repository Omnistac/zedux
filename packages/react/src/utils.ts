import { AnyAtomTemplate, Ecosystem } from '@zedux/atoms'
import React, { createContext } from 'react'

export const ecosystemContext = createContext('@@global')

/**
 * These are copied from packages/atoms/src/utils/general.ts
 *
 * IMPORTANT: keep these in-sync with the ones in the atoms package
 */
export const Explicit = 1
export const External = 2
export const Static = 4
// export const Deferred = 8

export const destroyed = Symbol.for(`@@zedux/destroyed`)

export const getReactContext = (
  ecosystem: Ecosystem,
  atom: AnyAtomTemplate
) => {
  const reactStorage: {
    contexts: WeakMap<typeof createContext, Record<string, React.Context<any>>>
  } = (ecosystem._storage.react ||= {})

  const contexts = (reactStorage.contexts ||= new WeakMap())
  const windowContexts =
    contexts.get(createContext) ||
    (contexts.set(createContext, {}).get(createContext) as Record<
      string,
      React.Context<any>
    >)

  return (windowContexts[atom.key] ||= createContext(
    undefined
  ) as React.Context<any>)
}
