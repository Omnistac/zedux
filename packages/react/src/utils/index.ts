export * from './general'
export * from './types'
