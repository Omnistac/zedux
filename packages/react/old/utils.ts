import { Atom, Scope } from './types'

let idCounter = 0

const generateId = () => {
  idCounter += 1
  return idCounter
}

const stringifyParam = (param: any): string => {
  if (Array.isArray(param)) {
    return `[${param.map(child => stringifyParam(child)).join(',')}]`
  }

  if (param && typeof param === 'object') {
    return `{${Object.entries(param)
      .map(([key, val]) => `${key}: ${stringifyParam(val)}`)
      .sort((a, b) => a.localeCompare(b))
      .join(',')}}`
  }

  return JSON.stringify(param)
}

export const generateAppId = () => `app-${generateId()}`
export const generateImplementationId = () => `im-${generateId()}`
export const generateInstanceId = () => `in-${generateId()}`
export const generateLocalId = () => `lo-${generateId()}`

export const getFullKey = (atom: Atom, params?: any[]) => {
  if (atom.scope === Scope.local) return generateLocalId()

  if (!params?.length) return atom.key

  return `${atom.key}---${stringifyParam(params)}`
}
