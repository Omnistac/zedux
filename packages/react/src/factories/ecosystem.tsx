import { Ecosystem } from '../classes/Ecosystem'
import { EcosystemConfig } from '../types'

export const ecosystem = (config: EcosystemConfig) => new Ecosystem(config)
