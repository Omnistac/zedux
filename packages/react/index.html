<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Zedux Dev</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/index.ts"></script>
    <script type="module">
      import { atom, useAtomState } from './src/index'
      import { createElement } from 'react'
      import { render } from 'react-dom'

      const greetingAtom = atom('greeting', 'Hello, World!')

      const App = () => {
        const [state, setState] = useAtomState(greetingAtom)

        return createElement('div', null,
          createElement('p', null, state),
          createElement('input', {
            onChange: event => setState(event.currentTarget.value),
            value: state
          })
        )
      }

      render(createElement(App), document.getElementById('root'))
    </script>
  </body>
</html>
