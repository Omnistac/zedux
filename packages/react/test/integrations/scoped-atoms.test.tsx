import {
  atom,
  AtomProvider,
  Ecosystem,
  inject,
  injectAtomValue,
  NodeOf,
  useAtomInstance,
  useAtomValue,
} from '@zedux/react'
import React, { createContext, useState } from 'react'
import { renderInEcosystem } from '../utils/renderInEcosystem'
import { ecosystem } from '../utils/ecosystem'
import { act } from '@testing-library/react'
import { expectTypeOf } from 'expect-type'
import { mockConsole } from '../utils/console'

describe('scoped atoms', () => {
  test('read provided atoms', async () => {
    const parentAtom = atom('parent', () => 1)

    const childAtom = atom('child', () => {
      const instance = inject(parentAtom)

      expectTypeOf(instance).toEqualTypeOf<NodeOf<typeof parentAtom>>()

      return instance
    })

    function Child() {
      const childVal = useAtomValue(childAtom)

      return <div data-testid="child">{childVal}</div>
    }

    function Parent() {
      const parentInstance = useAtomInstance(parentAtom)

      return (
        <AtomProvider instance={parentInstance}>
          <Child />
        </AtomProvider>
      )
    }

    const { findByTestId } = renderInEcosystem(<Parent />)
    const child = await findByTestId('child')

    expect(child).toHaveTextContent('1')

    act(() => {
      ecosystem.getNode(parentAtom).set(2)
    })

    expect(child).toHaveTextContent('2')
  })

  test('read provided React context values', async () => {
    const context = createContext(1)

    const childAtom = atom('child', () => {
      const value = inject(context)

      expectTypeOf(value).toEqualTypeOf<number>()

      return value
    })

    function Child() {
      const childVal = useAtomValue(childAtom)

      return <div data-testid="child">{childVal}</div>
    }

    function Parent() {
      const [state, setState] = useState(1)

      return (
        <context.Provider value={state}>
          <Child />
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          >
            Update
          </button>
        </context.Provider>
      )
    }

    const { findByTestId } = renderInEcosystem(<Parent />)
    const button = await findByTestId('button')
    const child = await findByTestId('child')

    expect(child).toHaveTextContent('1')

    act(() => {
      button.click()
    })

    expect(child).toHaveTextContent('2')
  })

  test('when a scoped atom is provided different values, creates a new instance with a scoped id', async () => {
    const calls: any[] = []
    const context = createContext('a')
    const parentAtom = atom('parent', (initialValue: number) => initialValue)

    const nestedContextualAtom = atom('nested', () => {
      const parentInstance = inject(parentAtom)

      return parentInstance.get()
    })

    const childAtom = atom(
      'child',
      () => {
        const reactValue = inject(context)
        const parentInstance = inject(parentAtom)
        const nestedContextualValue = injectAtomValue(nestedContextualAtom)

        expectTypeOf(reactValue).toEqualTypeOf<string>()
        expectTypeOf(parentInstance).toEqualTypeOf<NodeOf<typeof parentAtom>>()

        return `${reactValue}${parentInstance.get()} ${nestedContextualValue}`
      },
      { ttl: 0 }
    )

    function Child() {
      const childVal = useAtomValue(childAtom)

      calls.push(childVal)

      return <div data-testid="child">{childVal}</div>
    }

    function Parent() {
      const [state1, setState1] = useState('a')
      const instance1 = useAtomInstance(parentAtom, [1])
      const instance2 = useAtomInstance(parentAtom, [2])

      // 4 pairs: context1+atom1, context1+atom2, context2+atom1, context2+atom2
      return (
        <>
          {[state1, 'b'].map((value, index) => (
            <context.Provider key={index} value={value}>
              {[instance1, instance2].map(instance => (
                <AtomProvider key={instance.id} instance={instance}>
                  <Child />
                </AtomProvider>
              ))}
            </context.Provider>
          ))}
          <button
            data-testid="button1"
            onClick={() => setState1(state => state + 'a')}
          >
            Update 1
          </button>
        </>
      )
    }

    const { findAllByTestId, findByTestId } = renderInEcosystem(<Parent />)
    const button1 = await findByTestId('button1')
    const children = await findAllByTestId('child')

    expect(children[0]).toHaveTextContent('a1 1')
    expect(children[1]).toHaveTextContent('a2 2')
    expect(children[2]).toHaveTextContent('b1 1')
    expect(children[3]).toHaveTextContent('b2 2')

    expect(calls).toEqual(['a1 1', 'a2 2', 'b1 1', 'b2 2'])
    calls.splice(0, calls.length)

    // 2 parents, 4 children, 2 nested, 6 external
    expect(ecosystem.n.size).toBe(14)

    expect(Object.keys(ecosystem.findAll())).toEqual([
      'child-@scope("a",parent-[1])',
      'child-@scope("a",parent-[2])',
      'child-@scope("b",parent-[1])',
      'child-@scope("b",parent-[2])',
      'nested-@scope(parent-[1])',
      'nested-@scope(parent-[2])',
      'parent-[1]',
      'parent-[2]',
    ])

    act(() => {
      button1.click()
    })

    expect(children[0]).toHaveTextContent('aa1 1')
    expect(children[1]).toHaveTextContent('aa2 2')
    expect(children[2]).toHaveTextContent('b1 1')
    expect(children[3]).toHaveTextContent('b2 2')

    expect(calls).toEqual(['aa1 1', 'aa2 2', 'b1 1', 'b2 2'])
    calls.splice(0, calls.length)

    expect(ecosystem.n.size).toBe(14) // 2 new children, 2 destroyed

    expect(Object.keys(ecosystem.findAll())).toEqual([
      'child-@scope("aa",parent-[1])',
      'child-@scope("aa",parent-[2])',
      'child-@scope("b",parent-[1])',
      'child-@scope("b",parent-[2])',
      'nested-@scope(parent-[1])',
      'nested-@scope(parent-[2])',
      'parent-[1]',
      'parent-[2]',
    ])

    act(() => {
      ecosystem.getNode(parentAtom, [1]).set(11)
    })

    expect(children[0]).toHaveTextContent('aa11 11')
    expect(children[1]).toHaveTextContent('aa2 2')
    expect(children[2]).toHaveTextContent('b11 11')
    expect(children[3]).toHaveTextContent('b2 2')

    // atoms = fewer rerenders :muscle:
    expect(calls).toEqual(['aa11 11', 'b11 11'])
    calls.splice(0, calls.length)

    expect(ecosystem.n.size).toBe(14) // no changes

    expect(Object.keys(ecosystem.findAll())).toEqual([
      'child-@scope("aa",parent-[1])',
      'child-@scope("aa",parent-[2])',
      'child-@scope("b",parent-[1])',
      'child-@scope("b",parent-[2])',
      'nested-@scope(parent-[1])',
      'nested-@scope(parent-[2])',
      'parent-[1]',
      'parent-[2]',
    ])
  })

  test('scoped nodes propagate scope recursively to all observers', async () => {
    const calls: any[] = []
    const context = createContext('a')
    const parentAtom1 = atom('parent1', (initialValue: number) => initialValue)
    const parentAtom2 = atom('parent2', (initialValue: number) => initialValue)

    const topSource = atom(
      'topSource',
      () => {
        const parentInstance1 = inject(parentAtom1)
        const reactValue = inject(context)

        return reactValue + parentInstance1.get()
      },
      { ttl: 0 }
    )

    const middleSource = atom(
      'middleSource',
      () => {
        const parentInstance2 = inject(parentAtom2)

        return `${injectAtomValue(topSource)} ${parentInstance2.get()}`
      },
      { ttl: 0 }
    )

    const bottomSource = ({ get }: Ecosystem) => get(middleSource)

    const childAtom = atom(
      'child',
      () => {
        const value = injectAtomValue(bottomSource)

        expectTypeOf(value).toEqualTypeOf<string>()
        calls.push(value)

        return value
      },
      { ttl: 0 }
    )

    function Child() {
      const childVal = useAtomValue(childAtom)

      return <div data-testid="child">{childVal}</div>
    }

    function Parent() {
      const [state1, setState1] = useState('a')
      const instance1 = useAtomInstance(parentAtom1, [1])
      const instance100 = useAtomInstance(parentAtom1, [100])
      const instance2 = useAtomInstance(parentAtom2, [2])
      const instance200 = useAtomInstance(parentAtom2, [200])

      // 8 pairs
      return (
        <>
          {[state1, 'b'].map((value, index) => (
            <context.Provider key={index} value={value}>
              {[
                [instance1, instance2],
                [instance100, instance200],
              ].map(([instanceA, instanceB]) => (
                <AtomProvider key={instanceA.id} instance={instanceA}>
                  <AtomProvider key={instanceB.id} instance={instanceB}>
                    <Child />
                  </AtomProvider>
                </AtomProvider>
              ))}
            </context.Provider>
          ))}
          <button
            data-testid="button1"
            onClick={() => setState1(state => state + 'a')}
          >
            Update 1
          </button>
        </>
      )
    }

    const { findAllByTestId, findByTestId } = renderInEcosystem(<Parent />)
    const button1 = await findByTestId('button1')
    const children = await findAllByTestId('child')

    expect(children[0]).toHaveTextContent('a1 2')
    expect(children[1]).toHaveTextContent('a100 200')
    expect(children[2]).toHaveTextContent('b1 2')
    expect(children[3]).toHaveTextContent('b100 200')

    expect(calls).toEqual(['a1 2', 'a100 200', 'b1 2', 'b100 200'])
    calls.splice(0, calls.length)

    // 4 parents, 4 children, 4 nested, 4 middle, 4 top, 8 external
    expect(ecosystem.n.size).toBe(28)

    expect(Object.keys(ecosystem.findAll())).toMatchSnapshot()

    act(() => {
      button1.click()
    })

    expect(children[0]).toHaveTextContent('aa1 2')
    expect(children[1]).toHaveTextContent('aa100 200')
    expect(children[2]).toHaveTextContent('b1 2')
    expect(children[3]).toHaveTextContent('b100 200')

    expect(calls).toEqual(['aa1 2', 'aa100 200'])
    calls.splice(0, calls.length)

    // 4 parents, 4 children, 4 nested, 4 middle, 4 top, 8 external
    expect(ecosystem.n.size).toBe(28)

    expect(Object.keys(ecosystem.findAll())).toMatchSnapshot()

    const scope = new Map<Record<string, any>, any>([
      [context, 'b'],
      [parentAtom1, ecosystem.getNode(parentAtom1, [1])],
      [parentAtom2, ecosystem.getNode(parentAtom2, [2])],
    ])

    const childInstance = ecosystem.withScope(scope, () =>
      ecosystem.getNode(childAtom)
    )

    act(() => {
      childInstance.invalidate()
    })

    expect(calls).toEqual(['b1 2'])
    calls.splice(0, calls.length)

    // 4 parents, 4 children, 4 nested, 4 middle, 4 top, 8 external
    expect(ecosystem.n.size).toBe(28)

    expect(Object.keys(ecosystem.findAll())).toMatchSnapshot()

    act(() => {
      childInstance.destroy(true)
    })

    expect(calls).toEqual(['b1 2'])
    calls.splice(0, calls.length)

    // 4 parents, 4 children, 4 nested, 4 middle, 4 top, 8 external
    expect(ecosystem.n.size).toBe(28)

    expect(Object.keys(ecosystem.findAll())).toMatchSnapshot()
  })

  test('React context reference changes create new scopes', async () => {
    const calls: any[] = []
    const context = createContext({ a: { b: 1 } })

    const childAtom = atom('child', () => {
      const value = inject(context)

      expectTypeOf(value).toEqualTypeOf<{ a: { b: number } }>()
      calls.push(value)

      return value
    }) // no ttl

    function Child() {
      const childVal = useAtomValue(({ get }) => get(childAtom))

      return <div data-testid="child">{childVal.a.b}</div>
    }

    function Parent() {
      const [state, setState] = useState(1)

      return (
        <context.Provider value={{ a: { b: state } }}>
          <Child />
          <button
            data-testid="button"
            onClick={() => setState(state => state + 1)}
          >
            Update
          </button>
        </context.Provider>
      )
    }

    const { findByTestId } = renderInEcosystem(<Parent />)
    const button = await findByTestId('button')
    const child = await findByTestId('child')

    expect(child).toHaveTextContent('1')
    expect(calls).toEqual([{ a: { b: 1 } }])
    calls.splice(0, calls.length)

    expect(Object.keys(ecosystem.findAll())).toEqual([
      '@@selector-unnamed-0-@scope({"a":{"b":1}})',
      'child-@scope({"a":{"b":1}})',
    ])

    act(() => {
      button.click()
    })

    expect(child).toHaveTextContent('2')
    expect(calls).toEqual([{ a: { b: 2 } }])

    expect(Object.keys(ecosystem.findAll())).toEqual([
      '@@selector-unnamed-0-@scope({"a":{"b":1}})',
      'child-@scope({"a":{"b":1}})',
      'child-@scope({"a":{"b":2}})',
    ])
  })

  test('when a value is not provided, the thrown error attaches the requested context object', async () => {
    const consoleMock = mockConsole('error')
    const context = createContext<undefined | string>(undefined)
    const atom1 = atom('1', () => 'a')
    const childAtom1 = atom('child', () => inject(context))
    const childAtom2 = atom('child', () => inject(atom1))

    function Child1() {
      const childVal = useAtomValue(childAtom1)

      return <div data-testid="child">{childVal}</div>
    }

    function Child2() {
      const childVal = useAtomValue(childAtom2)

      return <div data-testid="child">{childVal}</div>
    }

    let error: Error | undefined = undefined

    try {
      renderInEcosystem(<Child1 />)
    } catch (err) {
      error = err as Error
    }

    expect(error?.message).toMatch(
      /Value was not provided to scoped atom. See attached cause/
    )
    expect(error?.cause).toBe(context)
    expect(consoleMock).toHaveBeenCalledTimes(2) // both Zedux and React log it

    try {
      renderInEcosystem(<Child2 />)
    } catch (err) {
      error = err as Error
    }

    expect(error?.message).toMatch(
      /Value was not provided to scoped atom. See attached cause/
    )
    expect(error?.cause).toBe(atom1)
    expect(consoleMock).toHaveBeenCalledTimes(4)
  })

  test('scoped atoms throw an error when retrieved outside a scoped context', () => {
    const consoleMock = mockConsole('error')
    const atom1 = atom('1', () => 'a')
    const childAtom = atom('child', () => inject(atom1))

    expect(() => ecosystem.getNode(childAtom)).toThrowError(
      /Scoped atom was initialized outside a scoped context/
    )
    expect(consoleMock).toHaveBeenCalledTimes(1)
  })

  test('scopes are recursive', () => {
    const context1 = atom('context1', () => 'a')
    const context2 = atom('context2', () => 'b')
    const context3 = atom('context3', () => 'c')

    const scope1 = [ecosystem.getNode(context1)]
    const scope2 = new Map([[context2, ecosystem.getNode(context2)]])
    const scope3 = [ecosystem.getNode(context3)]

    const scoped1 = atom('scoped1', () => inject(context1))
    const scoped2 = atom('scoped2', () => inject(context2))
    const scoped3 = atom('scoped3', () => inject(context3))

    const result1 = ecosystem.withScope(scope1, () => {
      const node1 = ecosystem.getNode(scoped1)

      const result2 = ecosystem.withScope(scope2, () => {
        const node1 = ecosystem.getNode(scoped1) // looks in outer scope
        const node2 = ecosystem.getNode(scoped2)

        const result3 = ecosystem.withScope(scope3, () => {
          const node1 = ecosystem.getNode(scoped1) // looks in outer outer scope
          const node2 = ecosystem.getNode(scoped2) // looks in outer scope
          const node3 = ecosystem.getNode(scoped3)

          return node1.get() + node2.get() + node3.get()
        })

        return `${node1.get() + node2.get()} ${result3}`
      })

      return `${node1.get()} ${result2}`
    })

    expect(result1).toBe('a ab abc')
  })

  test('scoped atoms are not hydrated', () => {
    const calls: any[] = []
    const context1 = atom('context1', () => 'a')
    const scope1 = [ecosystem.getNode(context1)]

    const scoped1 = atom('scoped1', () => {
      const val = inject(context1).get()

      calls.push(val)

      return val
    })

    const expectedId = 'scoped1-@scope(context1)'
    const hydration = { [expectedId]: 'b' }

    ecosystem.hydrate(hydration)

    const scopedNode1 = ecosystem.withScope(scope1, () =>
      ecosystem.getNode(scoped1)
    )

    expect(scopedNode1.get()).toBe('a')
    expect(ecosystem.hydration).toEqual(hydration)
    expect(scopedNode1.id).toBe(expectedId)
    expect(calls).toEqual(['a'])
  })

  test('ecosystem cycle event listeners receive the fully-scoped id when the node becomes Active', () => {
    jest.useFakeTimers()
    const calls: any[] = []
    const context1 = atom('context1', () => 'a')
    const scope1 = [ecosystem.getNode(context1)]

    const scoped1 = atom('scoped1', () => {
      const val = inject(context1).get()

      calls.push(val)

      return val
    })

    const expectedId = 'scoped1-@scope(context1)'

    ecosystem.on('cycle', event => {
      calls.push([event.source?.id, event.newStatus])
    })

    const scopedNode1 = ecosystem.withScope(scope1, () =>
      ecosystem.getNode(scoped1)
    )

    jest.runAllTimers()

    expect(scopedNode1.get()).toBe('a')
    expect(calls).toEqual(['a', [expectedId, 'Active']])
  })
})
