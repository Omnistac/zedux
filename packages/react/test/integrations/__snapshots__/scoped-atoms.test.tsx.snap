// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scoped atoms scoped nodes propagate scope recursively to all observers 1`] = `
[
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"a")",
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"b")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"a")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"b")",
  "child-@scope(parent2-[2],parent1-[1],"a")",
  "child-@scope(parent2-[2],parent1-[1],"b")",
  "child-@scope(parent2-[200],parent1-[100],"a")",
  "child-@scope(parent2-[200],parent1-[100],"b")",
  "middleSource-@scope(parent2-[2],parent1-[1],"a")",
  "middleSource-@scope(parent2-[2],parent1-[1],"b")",
  "middleSource-@scope(parent2-[200],parent1-[100],"a")",
  "middleSource-@scope(parent2-[200],parent1-[100],"b")",
  "parent1-[1]",
  "parent1-[100]",
  "parent2-[2]",
  "parent2-[200]",
  "topSource-@scope(parent1-[1],"a")",
  "topSource-@scope(parent1-[1],"b")",
  "topSource-@scope(parent1-[100],"a")",
  "topSource-@scope(parent1-[100],"b")",
]
`;

exports[`scoped atoms scoped nodes propagate scope recursively to all observers 2`] = `
[
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"b")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"b")",
  "child-@scope(parent2-[2],parent1-[1],"aa")",
  "child-@scope(parent2-[2],parent1-[1],"b")",
  "child-@scope(parent2-[200],parent1-[100],"aa")",
  "child-@scope(parent2-[200],parent1-[100],"b")",
  "middleSource-@scope(parent2-[2],parent1-[1],"aa")",
  "middleSource-@scope(parent2-[2],parent1-[1],"b")",
  "middleSource-@scope(parent2-[200],parent1-[100],"aa")",
  "middleSource-@scope(parent2-[200],parent1-[100],"b")",
  "parent1-[1]",
  "parent1-[100]",
  "parent2-[2]",
  "parent2-[200]",
  "topSource-@scope(parent1-[1],"aa")",
  "topSource-@scope(parent1-[1],"b")",
  "topSource-@scope(parent1-[100],"aa")",
  "topSource-@scope(parent1-[100],"b")",
]
`;

exports[`scoped atoms scoped nodes propagate scope recursively to all observers 3`] = `
[
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"b")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"b")",
  "child-@scope(parent2-[2],parent1-[1],"aa")",
  "child-@scope(parent2-[2],parent1-[1],"b")",
  "child-@scope(parent2-[200],parent1-[100],"aa")",
  "child-@scope(parent2-[200],parent1-[100],"b")",
  "middleSource-@scope(parent2-[2],parent1-[1],"aa")",
  "middleSource-@scope(parent2-[2],parent1-[1],"b")",
  "middleSource-@scope(parent2-[200],parent1-[100],"aa")",
  "middleSource-@scope(parent2-[200],parent1-[100],"b")",
  "parent1-[1]",
  "parent1-[100]",
  "parent2-[2]",
  "parent2-[200]",
  "topSource-@scope(parent1-[1],"aa")",
  "topSource-@scope(parent1-[1],"b")",
  "topSource-@scope(parent1-[100],"aa")",
  "topSource-@scope(parent1-[100],"b")",
]
`;

exports[`scoped atoms scoped nodes propagate scope recursively to all observers 4`] = `
[
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[2],parent1-[1],"b")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"aa")",
  "@@selector-bottomSource-0-@scope(parent2-[200],parent1-[100],"b")",
  "child-@scope(parent2-[2],parent1-[1],"aa")",
  "child-@scope(parent2-[2],parent1-[1],"b")",
  "child-@scope(parent2-[200],parent1-[100],"aa")",
  "child-@scope(parent2-[200],parent1-[100],"b")",
  "middleSource-@scope(parent2-[2],parent1-[1],"aa")",
  "middleSource-@scope(parent2-[2],parent1-[1],"b")",
  "middleSource-@scope(parent2-[200],parent1-[100],"aa")",
  "middleSource-@scope(parent2-[200],parent1-[100],"b")",
  "parent1-[1]",
  "parent1-[100]",
  "parent2-[2]",
  "parent2-[200]",
  "topSource-@scope(parent1-[1],"aa")",
  "topSource-@scope(parent1-[1],"b")",
  "topSource-@scope(parent1-[100],"aa")",
  "topSource-@scope(parent1-[100],"b")",
]
`;
